README.md
flutter\_spinning\_wheel [#](#flutter_spinning_wheel)
=====================================================

[![Build Status](https://travis-ci.com/davidanaya/flutter-spinning-wheel.svg?branch=master)](https://travis-ci.com/davidanaya/flutter-spinning-wheel) [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT) [![Pub](https://img.shields.io/pub/v/flutter_spinning_wheel.svg)](https://pub.dartlang.org/packages/flutter_spinning_wheel)

A customizable widget to use as a spinning wheel in Flutter.

Getting Started [#](#getting-started)
-------------------------------------

*   [Installation](#installation)
*   [Basic Usage](#basic-usage)
*   [Use Cases](#use-cases)


 <img  src="https://raw.githubusercontent.com/cocos3x/wheel_spin/main/assets/screenshot.jpg"  width="500px">

### Installation 


Add

    
    flutter_spinning_wheel : ^lastest_version
    
    



to your pubspec.yaml, and run

    flutter packages get
    



in your project's root directory.

### Basic Usage

Create a new project with command

    flutter create myapp
    



Edit lib/main.dart like this:

    class MyApp extends StatelessWidget {
    const MyApp({Key? key}) : super(key: key);

    // This widget is the root of your application.
    @override
    Widget build(BuildContext context) {
        return MaterialApp(
        title: 'Flutter Demo',
        theme: ThemeData(
            primarySwatch: Colors.blue,
        ),
        home: const MyHomePage(title: 'Flutter Demo Home Page'),
        );
    }
    }

    class MyHomePage extends StatefulWidget {
    const MyHomePage({Key? key, required this.title}) : super(key: key);


    final String title;

    @override
    State<MyHomePage> createState() => _MyHomePageState();
    }

    class _MyHomePageState extends State<MyHomePage> {
    
    WheelSpinController wheelSpinController = WheelSpinController();
    @override
    Widget build(BuildContext context) {
        // This method is rerun every time setState is called, for instance as done
        // by the _incrementCounter method above.
        //
        // The Flutter framework has been optimized to make rerunning build methods
        // fast, so that you can just rebuild anything that needs updating rather
        // than having to individually change instances of widgets.
        return Scaffold(
            appBar: AppBar(
            // Here we take the value from the MyHomePage object that was created by
            // the App.build method, and use it to set our appbar title.
            title: Text(widget.title),
            ),
            body: Center(
            // Center is a layout widget. It takes a single child and positions it
            // in the middle of the parent.
            child: Column(
                children: [
                WheelSpin(
                    controller: wheelSpinController,
                    pathImage: 'assets/wheel.png',
                    withWheel: 300,
                    pieces: 6,
                    // speed: 500,//defuaft is 500
                    isShowTextTest: true,
                    // offset: 1 / (10 * 6), //random 1/10 pieces defuaft is zero
                ),
                TextButton(
                    onPressed: () {
                        wheelSpinController.startWheel();
                    },
                    child: Text("Start")),
                TextButton(
                    onPressed: () {
                        wheelSpinController.stopWheel(2);
                    },
                    child: Text("Stop"))
                ],
            ),
            ));
    }
    }



You can replace the image with one of your preference.



### Use Cases

#### Game

    import 'dart:async';
    import 'dart:math';
    
    import 'package:flutter/material.dart';
    import 'package:flutter/services.dart';
    import 'package:flutter_spinning_wheel/flutter_spinning_wheel.dart';
    
    void main() {
      SystemChrome.setEnabledSystemUIOverlays([]);
      runApp(MyApp());
    }
    
    class MyApp extends StatelessWidget {
      @override
      Widget build(BuildContext context) {
        return MaterialApp(
          theme: ThemeData(
            primarySwatch: Colors.blue,
          ),
          home: MyHomePage(),
        );
      }
    }
    
    class MyHomePage extends StatelessWidget {
      final StreamController _dividerController = StreamController<int>();
    
      dispose() {
        _dividerController.close();
      }
    
      @override
      Widget build(BuildContext context) {
        return Scaffold(
          appBar: AppBar(backgroundColor: Color(0xffDDC3FF), elevation: 0.0),
          backgroundColor: Color(0xffDDC3FF),
          body: Center(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                SpinningWheel(
                  Image.asset('assets/images/roulette-8-300.png'),
                  width: 310,
                  height: 310,
                  initialSpinAngle: _generateRandomAngle(),
                  spinResistance: 0.6,
                  canInteractWhileSpinning: false,
                  dividers: 8,
                  onUpdate: _dividerController.add,
                  onEnd: _dividerController.add,
                  secondaryImage:
                      Image.asset('assets/images/roulette-center-300.png'),
                  secondaryImageHeight: 110,
                  secondaryImageWidth: 110,
                ),
                SizedBox(height: 30),
                StreamBuilder(
                  stream: _dividerController.stream,
                  builder: (context, snapshot) =>
                      snapshot.hasData ? RouletteScore(snapshot.data) : Container(),
                )
              ],
            ),
          ),
        );
      }
    
      double _generateRandomAngle() => Random().nextDouble() * pi * 2;
    }
    
   
    





CHANGELOG.md
first version : v1.0.0

test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:wheel_spin/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/wheelspin.dart
import 'dart:math';

import 'package:flutter/material.dart';

class WheelSpinController {
  late void Function() startWheel;
  late void Function(int index) stopWheel;
}

class WheelSpin extends StatefulWidget {
  final WheelSpinController controller;
  final String pathImage;
  final double withWheel;
  final int pieces;
  final double offset;
  final bool isShowTextTest;
  final int speed;
  const WheelSpin(
      {Key? key,
      required this.controller,
      required this.pathImage,
      required this.withWheel,
      this.offset = 0,
      required this.pieces,
      this.isShowTextTest = false,
      this.speed = 500})
      : super(key: key);

  @override
  State<WheelSpin> createState() => _WheelSpinState(controller);
}

class _WheelSpinState extends State<WheelSpin> with TickerProviderStateMixin {
  _WheelSpinState(WheelSpinController _controller) {
    _controller.startWheel = startWheel;
    _controller.stopWheel = stopWheel;
  }

  late final AnimationController _controllerStart = AnimationController(
    duration: Duration(milliseconds: widget.speed),
    vsync: this,
  );
  late final AnimationController _controllerFinish = AnimationController(
    duration: Duration(milliseconds: widget.speed * 2),
    vsync: this,
  );

  late final AnimationController _controllerMiddle = AnimationController(
    duration: Duration(milliseconds: widget.speed),
    vsync: this,
  );

  late final Animation<double> _animationFinish = CurvedAnimation(
    parent: _controllerFinish,
    curve: Curves.decelerate,
  );
  late final Animation<double> _animationStart = CurvedAnimation(
    parent: _controllerStart,
    curve: Curves.linear,
  );

  late final Animation<double> _animationMiddle = CurvedAnimation(
    parent: _controllerMiddle,
    curve: Curves.linear,
  );

  int statusWheel = 0;

  bool isNhanKetQua = false;
  int indexResuft = 0;
  double angle = 0;
  int pieces = 0;

  bool isStart = false;
  @override
  void initState() {
    super.initState();
    pieces = widget.pieces;

    _controllerStart.addStatusListener((status) {
      if (!isStart) return;
      print("status" + status.toString());
      if (status == AnimationStatus.completed) {
        if (!isNhanKetQua) {
          _controllerStart.reset();
          _controllerStart.forward();
        } else {
          setState(() {
            statusWheel = 1;
            _controllerStart.stop();

            _controllerMiddle.forward();
          });
        }
      }
    });
    _controllerMiddle.addListener(() {
      if (!isStart) return;
      double radiaus = indexResuft / pieces + widget.offset;
      if (_controllerMiddle.value >= radiaus) {
        setState(() {
          statusWheel = 2;
          angle = radiaus * 2 * pi;
          _controllerMiddle.stop();
          _controllerFinish.forward();
        });
      }
    });
  }

  reset() {
    setState(() {});
    isStart = false;
    statusWheel = 0;
    angle = 0;
    _controllerMiddle.reset();
    _controllerFinish.reset();
    _controllerStart.reset();

    isNhanKetQua = false;
  }

  void nhanketqua(int index) {
    isNhanKetQua = true;
    indexResuft = index;
  }

  Animation<double> getAnimation() {
    if (statusWheel == 0) return _animationStart;
    if (statusWheel == 1) return _animationMiddle;
    return _animationFinish;
  }

  @override
  void dispose() {
    if (_controllerStart.isAnimating) _controllerStart.dispose();
    if (_controllerFinish.isAnimating) _controllerFinish.dispose();
    if (_controllerMiddle.isAnimating) _controllerMiddle.dispose();
    super.dispose();
  }

  void startWheel() {
    reset();
    isStart = true;
    _controllerStart.forward();
  }

  void stopWheel(int index) {
    nhanketqua(index);
  }

  createLabels() {
    List<Widget> lsText = [];
    for (int i = 0; i < pieces; i++) {
      lsText.add(Transform.rotate(
        angle: -i * pi * 2 / widget.pieces,
        child: SizedBox(
          width: widget.withWheel,
          height: widget.withWheel,
          child: Stack(
            children: [
              Positioned(
                  left: widget.withWheel / 2,
                  top: widget.withWheel / 6,
                  child: Text(i.toString())),
            ],
          ),
        ),
      ));
    }
    return SizedBox(
      width: widget.withWheel,
      height: widget.withWheel,
      child: Center(
        child: Stack(
          children: lsText,
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return RotationTransition(
      turns: getAnimation(),
      child: Padding(
        padding: EdgeInsets.all(8.0),
        child: Transform.rotate(
          angle: angle,
          child: Stack(
            children: [
              Image.asset(
                widget.pathImage,
                fit: BoxFit.cover,
                width: widget.withWheel,
              ),
              Visibility(visible: widget.isShowTextTest, child: createLabels()),
            ],
          ),
        ),
      ),
    );
  }
}


windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.CreateAndShow(L"wheel_spin", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


lib/main.dart
import 'package:flutter/material.dart';
import 'package:wheel_spin/wheelspin.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  WheelSpinController wheelSpinController = WheelSpinController();
  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
        appBar: AppBar(
          // Here we take the value from the MyHomePage object that was created by
          // the App.build method, and use it to set our appbar title.
          title: Text(widget.title),
        ),
        body: Center(
          // Center is a layout widget. It takes a single child and positions it
          // in the middle of the parent.
          child: Column(
            children: [
              WheelSpin(
                controller: wheelSpinController,
                pathImage: 'assets/wheel.png',
                withWheel: 300,
                pieces: 6,
                // speed: 500,//defuaft is 500
                isShowTextTest: true,
                // offset: 1 / (10 * 6), //random 1/10 pieces defuaft is zero
              ),
              TextButton(
                  onPressed: () {
                    wheelSpinController.startWheel();
                  },
                  child: Text("Start")),
              TextButton(
                  onPressed: () {
                    wheelSpinController.stopWheel(2);
                  },
                  child: Text("Stop"))
            ],
          ),
        ));
  }
}
